### ISSUE TITLE
Implementing flash functionality in neovim

### ISSUE THREAD

**Issue by cuixiaorui** [VSCodeVim/Vim#8567]
I recently implemented flash-like plugin functionality in vscodeVim

I'll synchronize my progress in this issue

I also hope to share with you some ideas on how to make it work better.

**Comment by duanyongcheng77** [comment]
vscode vim 更新后是不是把你写的leap的功能给去掉了啊？

**Comment by cuixiaorui** [comment]
![flash-mvp-v1](https://github.com/VSCodeVim/Vim/assets/12064746/ddf82630-68a3-43a2-a363-2b6c3303c199)
Did a little more updating
Functions that have been realized now:
1. the most basic search function, press s and then input the characters to be searched.
2. can ignore case
3. support to use in visualization mode
4. support record jump point
5. displaying the entered search characters in statusBar

Basically, it's a working status. I've already used it in my daily work haha.

**Comment by aqav** [comment]
Good, I need it!

**Comment by sujianqingfeng** [comment]
great!

**Comment by aqav** [comment]
Would you create a pull request once you have done?

**Comment by macintacos** [comment]
Personally, I would be very frustrated if this only supported matching the start of words. Often I want to jump to the middle of a word, and doing that with Flash is very easy.

If the tool could be built to support only matching the beginning of words, I'd hope that would be a separate option, and not the default.

**Comment by theol0403** [comment]
Fwiw the real flash.nvim is supported with vscode-neovim.

**Comment by 42000** [comment]
Thank you very much! flash has to be the best plug-in by FAR

**Comment by EliTheSurfer** [comment]
Just wanted to say  thank you for this amazing work @cuixiaorui 👏

**Comment by armgilles** [comment]
Great work @cuixiaorui !

I think, it deserves **a merge of this feature**.

Perhaps, keep default key to `s` to don't change to much user habit (user can stil remap if he wants).;

**Comment by ibehnam** [comment]
Any updates on the merging of this? This is a great feature!

**Comment by ahmaddynugroho** [comment]
I'm moving to cvim

**Comment by cunbidun** [comment]
In the meantime, I implemented a minimal port of [flash.vscode](https://marketplace.visualstudio.com/items?itemName=cunbidun.flash-vscode) for personal use. It should works well with VSCodeVim.

**Comment by kik4444** [comment]
I appreciate all the work you've done to port this to vscodevim. However, I disagree with the choice to replace the native f. I often use f when I want to move a short distance, usually in the same line) to a unique character, or maybe a unique character several lines down. Then I only press f, the character I want, and I'm there. Using flash for this case means having to press enter in addition, which takes longer. And if I make a mistake, I can't press n or ; to move to the next match.

Coming from lazyvim, I've always used flash as a replacement for the normal search with / while f is normal, but can search across lines. In neovim, I usually press / to activate flash, then if I see the place I want to jump to on screen I just hit that button after typing in my search. Even before I hit the final button, my cursor is automatically moved to the first match. If the place I want is not visible on screen I can continue the search with n.

Would it be possible to implement that kind of functionality here?

**Comment by dautroc** [comment]
Hi, I've created an [extension](https://marketplace.visualstudio.com/items?itemName=dautroc.flash-vscode-dautroc) to simulate Flash. I would appreciate it if you could give it a try and provide me with your feedback. Thank you!

**Comment by Sourav9063** [comment]
# [Flash VsCode](https://marketplace.visualstudio.com/items?itemName=souravahmed.flash-vscode-latest)
[https://marketplace.visualstudio.com/items?itemName=souravahmed.flash-vscode-latest](https://marketplace.visualstudio.com/items?itemName=souravahmed.flash-vscode-latest)
## Table of Contents

- [Overview](#overview)
- [Installation](#installation)
- [Usage](#usage)
- [Configuration](#configuration)
- [Acknowledgements](#acknowledgements)

## Overview

**flash.vscode(latest)** is the easiest way to have the best plugin of Neovim, [flash.nvim](https://github.com/folke/flash.nvim) in Visual Studio Code. Works with [VScodeWeb](https://vscode.dev)

**You Don't Need To Know VIM to Use Flash VSCode**

## Installation

1. Open Visual Studio Code.
2. Go to the Extensions view (`Ctrl+Shift+X` on Windows/Linux or `Cmd+Shift+X` on macOS).
3. Search for **Flash VSCode**.
4. Click **Install**.
5. Reload VS Code if prompted.

## Usage

### Tutorial

[Tutorial Video](https://github.com/user-attachments/assets/b4660aa8-dd2d-4c9f-9622-c01521747a76)

1. **Activate Navigation:**
   Flash VSCode provides two main functionalities:

   - **`flash-vscode.start`**: `alt+f` Moves the cursor directly to the selected target.
   - `alt+f` while some text is selected will search for the selected text.

     ![flash jump](https://github.com/user-attachments/assets/9a416efd-0927-4df8-b1f1-81d1582f328c)

   - Press `alt+f` or `alt+shift+f` then `<search>` then `enter` to goto next match, `shift+enter` to goto previous match.
   - Press `alt+f` or `alt+shift+f` then `enter` to search previously entered query.
   - Select text and press `alt+f` or `alt+shift+f` then `enter` to search and mark the selected text.
   - Press `alt+f` or `alt+shift+f` then `ctrl+enter` to mark all the variables in the current file.

     ![flash enter](https://github.com/user-attachments/assets/e2f932e3-73c6-4acd-9d8c-9937bb116821)

   - Press `alt+j` or `alt+k` to mark all the next line or previous line.
     | Next Line | Previous Line |
     | :--------------------------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------------------------: |
     | ![Screenshot (178)](https://github.com/user-attachments/assets/9281233c-2021-4a4a-9a8b-e5e0bdfa350c) | ![Screenshot (176)](https://github.com/user-attachments/assets/ef55f28b-3560-4884-a131-b2ac04ec9453) |

2. **Selection:**

   - **`flash-vscode.startSelection`**: `alt+shift+f` Extends the selection from the original position to the target.

     ![flash select](https://github.com/user-attachments/assets/e3a12392-3ab5-4ff7-a657-f28c4b09da2d)

3. **Cancel Navigation:**
   - Press `Backspace` to remove the last character of your query, or press `Escape` to exit jump mode.

## Configuration

### Case Sensitivity

By default, `flash-vscode`'s search is smart case insensitive. If any uppercase latter exists then becomes case sensitive. To change this behavior, add to your settings:

```json
{
  "flash-vscode.caseSensitive": false
}
```

### Appearance Customization

The following configuration options allow you to customize the visual appearance of Flash VSCode:

```json
{
  "flash-vscode.dimOpacity": "0.65",
  "flash-vscode.matchColor": "#3e68d7",
  "flash-vscode.matchFontWeight": "bold",
  "flash-vscode.labelColor": "#ffffff",
  "flash-vscode.labelBackgroundColor": "#ff007c",
  "flash-vscode.labelQuestionBackgroundColor": "#3E68D7",
  "flash-vscode.labelFontWeight": "bold",
  "flash-vscode.labelKeys": "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()-_=+[]{}|;:'\",.<>/?"
}
```

- `flash-vscode.dimOpacity`: Opacity used to dim text.
- `flash-vscode.matchColor`: Color used for matched text.
- `flash-vscode.matchFontWeight`: Font weight for matched text.
- `flash-vscode.labelColor`: Color used for label text.
- `flash-vscode.labelBackgroundColor`: Background color for labels.
- `flash-vscode.labelQuestionBackgroundColor`: Background color for question labels.
- `flash-vscode.labelFontWeight`: Font weight for label text.
- `flash-vscode.labelKeys`: Characters to use for labels.

### VSCodeVim Integration (Optional)

To invoke Flash VSCode commands from VSCodeVim, in your `settings.json`, add entries to `"vim.normalModeKeyBindingsNonRecursive"` as follows:

```json
"vim.normalModeKeyBindingsNonRecursive": [
  {
    "before": ["s"],
    "commands": ["flash-vscode.start"]
  },
  {
    "before": ["S"],
    "commands": ["flash-vscode.startSelection"]
  },
  {
    "before": [ "<BS>" ],
    "commands": [ "flash-vscode.backspace" ]
  },
]
```

This configuration triggers Flash VSCode when you press `s` or `S` in normal mode.

## Acknowledgements

- [flash.nvim](https://github.com/folke/flash.nvim) for the original ideas.
- [Jumpy2](https://marketplace.visualstudio.com/items?itemName=DavidLGoldberg.jumpy2) for some of the implementation details.
- [flash.vscode](https://github.com/cunbidun/flash.vscode) flash.vscode(latest) extension is supper set of this extension.
- [CVim-PR](https://github.com/VSCodeVim/Vim/issues/8567) [CVim](https://github.com/cuixiaorui/vscodeVim/tree/flash) for ux improvement ideas.
